bson/src/main/org/bson/codecs/AbstractMapCodec.java

```diff
@@ -0,0 +1,124 @@
+
+import org.bson.BsonBinarySubType;
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.Transformer;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static java.util.Arrays.asList;
+import static org.bson.assertions.Assertions.notNull;
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+
+abstract class AbstractMapCodec<T extends Map<String, Object>>  implements Codec<T> {
+    private static final CodecRegistry DEFAULT_REGISTRY = fromProviders(asList(new ValueCodecProvider(), new BsonValueCodecProvider(),
+            new DocumentCodecProvider(), new IterableCodecProvider(), new MapCodecProvider()));
+    private static final BsonTypeClassMap DEFAULT_BSON_TYPE_CLASS_MAP = new BsonTypeClassMap();
+    private final BsonTypeCodecMap bsonTypeCodecMap;
+    private final CodecRegistry registry;
+    private final Transformer valueTransformer;
+
+    AbstractMapCodec() {
+        this(DEFAULT_REGISTRY);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry) {
+        this(registry, DEFAULT_BSON_TYPE_CLASS_MAP);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry, final BsonTypeClassMap bsonTypeClassMap) {
+        this(registry, DEFAULT_BSON_TYPE_CLASS_MAP, null);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry, final BsonTypeClassMap bsonTypeClassMap, final Transformer valueTransformer) {
+        this.registry = notNull("registry", registry);
+        this.bsonTypeCodecMap = new BsonTypeCodecMap(notNull("bsonTypeClassMap", bsonTypeClassMap), registry);
+        this.valueTransformer = valueTransformer != null ? valueTransformer : new Transformer() {
+            @Override
+            public Object transform(final Object value) {
+                return value;
+            }
+        };
+    }
+
+    abstract T newInstance();
+
+    abstract void beforeFields(BsonWriter bsonWriter, EncoderContext encoderContext, T map);
+
+    abstract boolean skipField(EncoderContext encoderContext, String key);
+
+    void encodeMap(final BsonWriter writer, final T map, final EncoderContext encoderContext) {
+        writeMap(writer, map, encoderContext);
+    }
+
+    T decodeMap(final BsonReader reader, final DecoderContext decoderContext) {
+        T map = newInstance();
+
+        reader.readStartDocument();
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            String fieldName = reader.readName();
+            map.put(fieldName, readValue(reader, decoderContext));
+        }
+
+        reader.readEndDocument();
+        return map;
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    void writeValue(final BsonWriter writer, final EncoderContext encoderContext, final Object value) {
+        if (value == null) {
+            writer.writeNull();
+        } else {
+            Codec codec = registry.get(value.getClass());
```

Comment:
I have a couple of concerns with this change:

1. This will incur a slight performance penalty for every nested array and map due to the extra lookup in the registry.  You should measure to see if it makes a difference.
2. More importantly, this will break existing applications which haven't added the new MapCodecProvider to the registry, or didn't bother to add the IterableCodecProvider; e.g. this will fail on top of this patch:

     collection.insertOne(new Document('nested', new HashMap<String, Object>()))

  where currently is succeeds (this case would be fixed by added MapCodecProvider to the default registry, but other applications may be replacing that registry with their own).